File: 94. Binary Tree Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal_withRecursion(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return None
            
        result=[]
        
        left=self.inorderTraversal(root.left)
        if left:
            result+=left
                
        result.append(root.val)
        
        right=self.inorderTraversal(root.right)
        if right:
            result+=right
                   
        return result



    def inorderTraversal_withoutRecursion(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return None
        
        cur=root
        S=Stack()
        result=[]
        
        while(cur):
            S.push(cur)        
            cur=cur.left
            
        cur=S.top()
        
        while(not S.is_empty()):
            cur=S.pop()
            
            if cur.right is None:
                result.append(cur.val)
            else:
                result.append(cur.val)
                cur=cur.right
                while(cur):
                    S.push(cur)
                    cur=cur.left
                    
        return result

class Stack():
    def __init__(self):
        self._data=[]

    def is_empty(self):
        return len(self._data)==0
    
    def push(self,e):
        self._data.append(e)
      
    def pop(self):
        if self.is_empty():
            return None
        return self._data.pop()
    
    def top(self):
        if self.is_empty():
            return None
        return self._data[-1]
    
    
    
    
    
    
    
    
    
    
    
    
        

File: 94.Binary_Tree_Inorder_Traversal.cpp
#include<vector>
#include<iostream>
#include<string>
using namespace std;

struct TreeNode{    
    int val;
    TreeNode *left=nullptr;
    TreeNode *right=nullptr;
};

vector<int> InorderTraversal(TreeNode *root);

int main(){
    TreeNode root, n1, n2, n3, n4, n5;
    root.val = 0, root.left = &n1, root.right = &n2;
    n1.val = 1, n1.left = &n3, n1.right = &n4;
    n2.val = 2;
    n3.val = 3, n3.right = &n5;
    n4.val = 4, n5.val = 5;

    auto ans = InorderTraversal(&root);
    for(auto i:ans){
        cout << i << ' ';
    }
    //cout << InorderTraversal(&root) << endl;
}

vector<int> InorderTraversal(TreeNode *root){
    
    if(root==nullptr){
        vector<int> res;
        return res;
        }

    vector<int> left = InorderTraversal(root->left);

    left.insert(left.end(), root->val);

    vector<int> right = InorderTraversal(root->right);

    if(!right.empty())
        left.insert(left.end(), right.begin(), right.end());

    return left;
}